### PR TITLE
new component selectors

### PR DESCRIPTION
**Pull request by yoshokatana**
* [trello ticket](https://trello.com/c/VC84WFok/323-component-selector-updates)
* new selector is smaller, shows outlines of component
* doesn't push text around when opening inline forms
* better action + info layout
* not confined to dimensions of component
* new icons for parent, settings, delete

**Adding Components**

* add component button is now in selector
* allows you to add components right after selected component (not just end of the list)
* selects components after you add them (more visual confirmation of your actions)
* placeholders for empty component lists now let you add components into them

**Focusing Forms**

* more robust caret setting when opening forms
* faster navigation between parent and child components

f5a3c2f4-e7a1-42e5-9470-a9b12aec5e96-476-0004af42b68f7feb

### CODE REVIEW

File: services/select.js [hunk 1]
```diff
@@ -208,193 +167,250 @@ function addIframeOverlays(el) {
...
-  var selectedLabel = componentBar.querySelector('.selected-label'),
-    dragIcon = dom.create(`<img src="${site.get('assetPath')}/media/components/clay-kiln/component-bar-drag.svg" alt="Drag" class="drag-icon"></span>`);
-
-  selectedLabel.setAttribute('title', `Drag to reorder: ${label(selectedLabel.getAttribute('title'))}`);
-  el.classList.add('drag');
-  selectedLabel.insertBefore(dragIcon, selectedLabel.firstChild);
+function addListPlaceholder(parent) {
+  var listDiv = dom.find(parent.listEl, '.component-list-inner'),
+    children = _.filter(listDiv.children, function (child) {
+      return !_.includes(['SCRIPT', 'STYLE'], child.tagName);
```
Comment (yoshokatana) [inline services/select.js]: note: if a component is added, and its list element is its root element (e.g. `<aside data-uri="source-links" data-editable="content"></aside>`), any scripts or styles will end up wrapped inside the div.

idea: update the wrapper in the `component-list` decorator to _not_ wrap those

File: services/select.js [hunk 2]
```diff
@@ -208,193 +167,250 @@ function addIframeOverlays(el) {
...
+
+  if (parent.isComponentList) {
+    button.classList.remove(hidden);
+    button.addEventListener('click', function () {
+      var confirm = window.confirm('Delete this component? This cannot be undone.'); // eslint-disable-line
+
+      if (confirm) {
+        return edit.removeFromParentList({el: el, ref: options.ref, parentField: parent.path, parentRef: parent.ref})
+          .then(function () {
+            addListPlaceholder(parent);
```
Comment (yoshokatana) [inline services/select.js]: make this a single-line function

File: decorators/component-list.js [hunk 3]
```diff
@@ -191,67 +139,24 @@ function when(el, options) {
...
-  var args = _.get(options, 'data._schema.' + references.componentListProperty),
-    pane, button, dropArea;
-
-  // if _componentList: true, make the args an object
-  args = _.isObject(args) ? args : {};
-
-  args.field = _.omit(options, 'data'); // add button needs ref and path.
+  var wrappableEls = _.filter(el.childNodes, function (child) {
+      // wrap everything that ISN'T the component selector
+      return child.nodeType !== 1 || !child.classList.contains('component-selector');
```
Comment (yoshokatana) [inline decorators/component-list.js]: todo: wrap anything that isn't component selector, scripts, or styles

File: decorators/focus.js [hunk 4]
```diff
@@ -33,8 +33,16 @@ function getClickOffset(e) {
     if (dom.find(parent, '.kiln-placeholder') || dom.find(parent, '.kiln-permanent-placeholder')) {
       return 0;
     }
+
+    // grab all the text that isn't in the selector
+    // e.g. regular textnodes, formatted stuff
+    _.each(parent.childNodes, function (node) {
+      // if it's a text node (type is 3) or ISN'T the component selector, grab the text
+      if (node.nodeType === 3 || !node.classList.contains('component-selector')) {
```
Comment (yoshokatana) [inline decorators/focus.js]: todo: make sure this also doesn't grab inline scripts and styles

File: decorators/placeholder.js [hunk 5]
```diff
@@ -110,6 +121,41 @@ function isComponentListEmpty(data) {
 }
 
 /**
+ * get list element, even when we're re-adding placeholders to empty lists
```
Comment (yoshokatana) [inline decorators/placeholder.js]: todo: explain why!

"when re-running placeholders after deleting all items from a component list, the `.component-list-inner` div has already been added by the `component-list` decorator. when we look for the list to add the placeholder, we need to take that into account"

File: services/add-component-handler.test.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
+var dirname = __dirname.split('/').pop(),
+  filename = __filename.split('/').pop().split('.').shift(),
+  lib = require('./add-component-handler'),
+  dom = require('@nymag/dom');
+
+describe(dirname, function () {
+  describe(filename, function () {
```
Comment (yoshokatana) [inline services/add-component-handler.test.js]: test the logic of the default exported function (`addComponentHandler()`) as well

File: services/add-component.test.js [hunk 7]
```diff
@@ -13,13 +15,15 @@ describe(dirname, function () {
...
+      sandbox.stub(progress);
+      sandbox.stub(focus);
     });
 
     afterEach(function () {
       sandbox.restore();
     });
 
-    it('creates component and adds it before the pane', function () {
+    it('creates component and adds it at the end', function () {
```
Comment (yoshokatana) [inline services/add-component.test.js]: todo: add tests for adding component after a different component

File: services/select.js [hunk 8]
```diff
@@ -208,193 +167,250 @@ function addIframeOverlays(el) {
...
 
-  if (hasSettings) {
-    el = dom.create(`<li class="settings label">
-      <img src="${site.get('assetPath')}/media/components/clay-kiln/component-bar-settings.svg" alt="Settings">
-      <span class="menu-item">Settings</span>
-    </li>`);
+/**
+ * get parent info, if it exists
+ * @param {Element} el current component
+ * @returns {Promise} w/ empty object OR el, ref, path, list (from the schema), listEl, and isComponentList (boolean)
```
Comment (yoshokatana) [inline services/select.js]: todo: remove this. it's obvious from the `return` statement

File: services/select.js [hunk 9]
```diff
@@ -208,193 +167,250 @@ function addIframeOverlays(el) {
...
-  var el = dom.create(`<li class="delete label">
-    <img src="${site.get('assetPath')}/media/components/clay-kiln/component-bar-delete.svg" alt="Delete">
-    <span class="menu-item">Delete</span>
-  </li>`);
-
-  el.addEventListener('click', function () {
-    return edit.removeFromParentList(opts)
-      .then(forms.close);
-  });
+function addDeleteHandler(selector, parent, el, options) {
```
Comment (yoshokatana) [inline services/select.js]: todo: add tests for the placeholder adding

File: services/select.js [hunk 10]
```diff
@@ -208,193 +167,250 @@ function addIframeOverlays(el) {
...
+
+    // add events to the component itself
+    // when the component is clicked, it should be selected
+    el.addEventListener('click', componentClickHandler.bind(null, el));
+
+    // make sure components are relatively positioned
+    el.classList.add('component-selector-wrapper');
+
+    // add selector to the component el
+    dom.prependChild(el, selector);
```
Comment (yoshokatana) [inline services/select.js]: todo: explain why (add selector behind the rest of the components' elements, so you can access them)

File: services/tpl.js [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+ * @param {string} selector
+ * @returns {Element}
+ */
+function getTemplate(selector) {
+  var template = dom.find(selector);
+
+  return document.importNode(template.content, true);
+}
+
+module.exports.get = getTemplate;
```
Comment (yoshokatana) [inline services/tpl.js]: todo: add some tests for `getTemplate`, just to prove it works on all supported browsers

**Comment by yoshokatana** [comment]
todo: the new services should maybe be written as es6 modules? @byronhulcher @jonwinton @davidandrewfrey thoughts?

File: services/select.js (hunk 1 above)
Comment (yoshokatana) [inline services/select.js]: done

File: services/select.js (hunk 2 above)
Comment (yoshokatana) [inline services/select.js]: addListPlaceholder(parent)

File: decorators/component-list.js (hunk 3 above)
Comment (yoshokatana) [inline decorators/component-list.js]: ✓ done

File: decorators/focus.js (hunk 4 above)
Comment (yoshokatana) [inline decorators/focus.js]: actually, they won't be affected anyway

File: decorators/placeholder.js (hunk 5 above)
Comment (yoshokatana) [inline decorators/placeholder.js]: ✓ done

File: services/add-component-handler.test.js (hunk 6 above)
Comment (yoshokatana) [inline services/add-component-handler.test.js]: doing this in a separate pr

File: services/add-component.test.js (hunk 7 above)
Comment (yoshokatana) [inline services/add-component.test.js]: ✗ low roi

File: services/select.js (hunk 8 above)
Comment (yoshokatana) [inline services/select.js]: ✓ done

File: services/select.js (hunk 9 above)
Comment (yoshokatana) [inline services/select.js]: will do in a new pr

File: services/select.js (hunk 10 above)
Comment (yoshokatana) [inline services/select.js]: ✓ done

File: services/tpl.js (hunk 11 above)
Comment (yoshokatana) [inline services/tpl.js]: ✗ low roi, all modern browsers support this

File: decorators/focus.js [hunk 12]
```diff
@@ -14,7 +13,8 @@ var _ = require('lodash'),
  * @returns {number}
  */
 function getClickOffset(e) {
-  let range, textNode, offset, parent, parentText, parentOffset;
+  var parentText = '';
```
Comment (jonwinton) [inline decorators/focus.js]: @yoshokatana why the mixing `var` and `let`?

Comment (yoshokatana) [inline decorators/focus.js]: good catch. fixing

**Comment by jonwinton** [comment]
👍

**Comment by byronhulcher** [comment]
Great changes + rigorous commenting 👍

**Comment by cruzanmo** [comment]
Yay the parent component is now visible! +1

**Pull request merged**